pkg/scheduler/controller/shoot/scheduler_control.go

```diff
@@ -175,61 +176,98 @@ func determineSeed(shoot *gardencorev1beta1.Shoot, seedLister gardencorelisters.
+	}
+	return matchingSeeds, nil
 }
 
-func filterSeedsMatchingSeedSelector(seedList []*gardencorev1beta1.Seed, labelSelector *metav1.LabelSelector, kind string) ([]*gardencorev1beta1.Seed, error) {
-	selector := &metav1.LabelSelector{}
-	if labelSelector != nil {
-		selector = labelSelector
+func filterSeedsMatchingSeedSelector(seedList []*gardencorev1beta1.Seed, seedSelector *gardencorev1beta1.SeedSelector, kind string) ([]*gardencorev1beta1.Seed, error) {
+	if seedSelector == nil || seedSelector.LabelSelector == nil {
+		return seedList, nil
 	}
-	seedSelector, err := metav1.LabelSelectorAsSelector(selector)
+	selector, err := metav1.LabelSelectorAsSelector(seedSelector.LabelSelector)
 	if err != nil {
-		return nil, fmt.Errorf("label selector conversion failed: %v for seedSelector: %v", *selector, err)
+		return nil, fmt.Errorf("label selector conversion failed: %v for seedSelector: %v", *seedSelector.LabelSelector, err)
 	}
 
 	var matchingSeeds []*gardencorev1beta1.Seed
 	for _, seed := range seedList {
-		if !seedSelector.Matches(labels.Set(seed.Labels)) {
-			continue
+		if selector.Matches(labels.Set(seed.Labels)) {
+			matchingSeeds = append(matchingSeeds, seed)
 		}
-
-		matchingSeeds = append(matchingSeeds, seed)
 	}
 
 	if len(matchingSeeds) == 0 {
-		return nil, fmt.Errorf("none out of the %d seeds has the matching labels required by seed selector of '%s' (selector: '%s')", len(seedList), kind, seedSelector.String())
+		return nil, fmt.Errorf("none out of the %d seeds has the matching labels required by seed selector of '%s' (selector: '%s')", len(seedList), kind, selector.String())
 	}
 	return matchingSeeds, nil
 }
 
-func getCandidates(shoot *gardencorev1beta1.Shoot, seedList []*gardencorev1beta1.Seed, strategy config.CandidateDeterminationStrategy) ([]*gardencorev1beta1.Seed, error) {
+func filterSeedsMatchingProviders(cloudProfile *gardencorev1beta1.CloudProfile, shoot *gardencorev1beta1.Shoot, seedList []*gardencorev1beta1.Seed) ([]*gardencorev1beta1.Seed, error) {
+	var possibleProviders []string
+	if cloudProfile.Spec.SeedSelector != nil {
+		possibleProviders = cloudProfile.Spec.SeedSelector.ProviderTypes
+	}
+
+	var matchingSeeds []*gardencorev1beta1.Seed
+	for _, seed := range seedList {
+		if matchProvider(seed.Spec.Provider.Type, shoot.Spec.Provider.Type, possibleProviders) {
+			matchingSeeds = append(matchingSeeds, seed)
+		}
+	}
+	return matchingSeeds, nil
+}
+
+func applyStrategy(shoot *gardencorev1beta1.Shoot, seedList []*gardencorev1beta1.Seed, strategy config.CandidateDeterminationStrategy) ([]*gardencorev1beta1.Seed, error) {
 	var candidates []*gardencorev1beta1.Seed
 
 	switch {
 	case shoot.Spec.Purpose != nil && *shoot.Spec.Purpose == gardencorev1beta1.ShootPurposeTesting:
 		candidates = determineCandidatesOfSameProvider(seedList, shoot)
 	case strategy == config.SameRegion:
 		candidates = determineCandidatesWithSameRegionStrategy(seedList, shoot)
+	case strategy == config.BestRegion:
```

Comment:
Currently there are three strategies
 - same region -> same cloud provider + same region name of Seed and Shoot
 - bestRegion -> same region, if no match: minimal distance
 - minimal distance -> determine minimal distance (levenshtein distance) across provider types

Why do we need the "bestRegion" strategy? I think it would be easier to just have "minimalDistance" that (like before) executes the "sameRegion" and if unsuccessful uses the levenshtein distance.

If "bestRegion" region is wanted, then I am missing unit test for it - even though it is just a combination of two strategies.

Additionally, the current behaviour of the "minimalDistance" has changed as "sameRegion" is not executed any more.

Reply:
a matching region always has minimal distance. Therefore is should always be preferred.
I've removed the BestRegion again

Reply:
> a matching region always has minimal distance. Therefore it should always be preferred.
but two cloud providers can name their regions identically. In this case I would prefer using the same cloud provider (what same region strategy does).

Reply:
I could incorporate the provider in the distance, so that same provider always is closer